docs/using-the-compiler.rst

```diff
@@ -0,0 +1,144 @@
+
+The example below is presented in a human-readable way. Properly formatted metadata
+should use quotes correctly, reduce whitespace to a minimum and sort the keys of all objects
+to arrive at a unique formatting.
+
+Comments are of course not permitted and used here only for explanatory purposes.
+
+Input Description
+-----------------
+
+The input description is language-specific and could change with each compiler version, but it
+should be backwards compatible if possible.
+
+    {
+      sources:
+      {
+        "abc": "contract b{}",
+        "def": {keccak: "0x123..."}, // source has to be retrieved by its hash
+        "dir/file.sol": "contract a {}"
+      },
+      settings:
+      {
+        remappings: [":g/dir"],
+        optimizer: {enabled: true, runs: 500},
+        compilationTarget: "myFile.sol:MyContract", // Can also be an array
+        // To be backwards compatible, use the full name of the contract in the output
+        // only if there are name clashes.
+        // If the full name was given as "compilationTargets", use the full name.
+        libraries: {
+          "def:MyLib": "0x123123..."
+        },
+        // The following can be used to restrict the fields the compiler will output.
+        outputSelection: {
+          // to be defined
+        }
+      }
+    }
+
+Metadata Output
+---------------
+
+Note that the actual bytecode is not part of the metadata because the hash
+of the metadata structure will be included in the bytecode itself.
+
+This requires the compiler to be able to compute the hash of its own binary,
+which requires it to be statically linked. The hash of the binary is not
+too important. It is much more important to have the commit hash because
+that can be used to query a location of the binary (and whether the version is
+"official") at a registry contract. 
+
+    {
+      version: "1",
+      language: "Solidity",
+      compiler: {
+        commit: "55db20e32c97098d13230ab7500758e8e3b31d64",
+        version: "soljson-2313-2016-12-12",
+        keccak: "0x123..."
+      },
+      sources:
+      {
+        "abc": "contract b{}",
+        "def": {keccak: "0x123..."}, // source has to be retrieved by its hash
+        "dir/file.sol": "contract a {}"
+      },
+      settings:
+      {
+        remappings: [":g/dir"],
+        optimizer: {enabled: true, runs: 500},
+        compilationTarget: "myFile.sol:MyContract",
+        // To be backwards compatible, use the full name of the contract in the output
+        // only if there are name clashes.
+        // If the full name was given as "compilationTargets", use the full name.
+        libraries: {
+          "def:MyLib": "0x123123..."
+        }
+      },
+      output:
+      {
+        abi: [ /* abi definition */ ],
+        userDocumentation: [ /* user documentation comments */ ],
+        developerDocumentation: [ /* developer documentation comments */ ],
+        natspec: [ /* natspec comments */ ]
+      }
```

Comment:
Proposed names for the rest:

- `abi`
- `methodIdentifiers` / `functionHashes`
- `evmAssembly`
- `evmBytecode`
- `evmRuntimeBytecode`
- `evmGasEstimates`
- `evmOpcodes`
- `interface` or `solidityInterface`
- `developerDocumentation`
- `userDocumenation`
- `natspec`
- `why3`

I'm not fully sure about source maps?